File: src/dotnet/APIView/APIView/Model/V2/ReviewLine.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using APIView.TreeToken;

namespace APIView.Model.V2
{
    /// <summary>
    /// Review line object corresponds to each line displayed on API review. If an empty line is required then add a review line object without any token.
    /// </summary>
    public class ReviewLine
    {
        /// <summary>
        /// LineId is only required if we need to support commenting on a line that contains this token. 
        /// Usually code line for documentation or just punctuation is not required to have lineId.lineId should be a unique value within
        /// the review token file to use it assign to review comments as well as navigation Id within the review page.        /// for e.g Azure.Core.HttpHeader.Common, azure.template.template_main
        /// </summary>
        public string LineId { get; set; }
        public string CrossLanguageId { get; set; }
        /// <summary>
        /// List of tokens that constructs a line in API review
        /// </summary>
        public List<ReviewToken> Tokens { get; set; } = [];
        /// <summary>
        /// Add any child lines as children. For e.g. all classes and namespace level methods are added as a children of namespace(module) level code line.
        /// Similarly all method level code lines are added as children of it's class code line.
        /// </summary>
        public List<ReviewLine> Children { get; set; } = [];
        /// <summary>
        /// This is set if API is marked as hidden
        /// </summary>
        public bool? IsHidden { get; set; }
        /// <summary>
        /// This is set if a line is end of context. For e.g. end of a class or name space line "}"
        /// </summary>
        public bool? IsContextEndLine { get; set; }
        /// <summary>
        /// This is to set a line as related to another line. So when a related line is hidden in node or tree view then current line will also be hidden
        /// for e.g. an attribute line or notation line will be set as related to that API or class line.
        /// </summary>
        public string RelatedToLine { get; set; }
        // Following properties are helper methods that's used to render review lines to UI required format.
        [JsonIgnore]
        public DiffKind DiffKind { get; set; } = DiffKind.NoneDiff;
        [JsonIgnore]
        public bool IsActiveRevisionLine = true;
        [JsonIgnore]
        public bool IsDocumentation => Tokens.Count > 0 && Tokens[0].IsDocumentation == true;
        [JsonIgnore]
        public bool IsEmpty => Tokens.Count == 0 || !Tokens.Any( t => t.SkipDiff != true);
        [JsonIgnore]
        public bool Processed { get; set; } = false;

        [JsonIgnore]
        public int Indent {  get; set; }
        [JsonIgnore]
        public ReviewLine parentLine { get; set; }
        public void AddToken(ReviewToken token)
        {
            Tokens.Add(token);
        }

        public void AppendApiTextToBuilder(StringBuilder sb, int indent = 0, bool skipDocs = true, int lineIndentSpaces = 4)
        {
            if (skipDocs && Tokens.Count > 0 && Tokens[0].IsDocumentation == true)
            {
                return;
            }

            //Add empty line in case of review line without tokens
            if (Tokens.Count == 0)
            {
                sb.Append(Environment.NewLine);
                return;
            }
            //Add spaces for indentation
            for (int i = 0; i < indent; i++)
            {
                for(int j = 0; j < lineIndentSpaces; j++)
                {
                    sb.Append(" ");
                }
            }
            //Process all tokens
            sb.Append(ToString(true));
            
            sb.Append(Environment.NewLine);
            foreach (var child in Children)
            {
                child.AppendApiTextToBuilder(sb, indent + 1, skipDocs, lineIndentSpaces);
            }
        }

        private string ToString(bool includeAllTokens)
        {
            var filterdTokens = includeAllTokens ? Tokens: Tokens.Where(x => x.SkipDiff != true);
            if (!filterdTokens.Any())
            {
                return string.Empty;
            }
            StringBuilder sb = new();
            bool spaceAdded = false;
            foreach (var token in filterdTokens)
            {
                sb.Append(token.HasPrefixSpace == true && !spaceAdded ? " " : string.Empty);
                sb.Append(token.Value);
                sb.Append(token.HasSuffixSpace == true ? " " : string.Empty);
                spaceAdded = token.HasSuffixSpace == true;
            }
            return sb.ToString();
        }

        
        public override string ToString()
        {
            return ToString(false);
        }

        public override bool Equals(object obj)
        {
            if(obj is ReviewLine other)
            {
                return ToString() == other.ToString();
            }
            return false;
        }

        public override int GetHashCode()
        {
            return ToString().GetHashCode();
        }

        public string GetTokenNodeIdHash(string parentNodeIdHash, int lineIndex)
        {
            var idPart = LineId;
            var token = Tokens.FirstOrDefault(t => t.RenderClasses.Count > 0);
            if (token != null)
            {
                idPart = $"{idPart}-{token.RenderClasses.First()}";
            }
            idPart = $"{idPart}-{lineIndex}-{DiffKind}";
            var hash = CreateHashFromString(idPart);
            return hash + parentNodeIdHash.Replace("nId", "").Replace("root", ""); // Append the parent node Id to ensure uniqueness
        }

        private static string CreateHashFromString(string inputString)
        {
            int hash = inputString.GetHashCode();
            return "nId" + hash.ToString();
        }
    }
}

File: src/java/apiview-java-processor/src/main/java/com/azure/tools/apiview/processor/model/ReviewLine.java
package com.azure.tools.apiview.processor.model;

import com.azure.json.JsonSerializable;
import com.azure.json.JsonWriter;
import com.azure.tools.apiview.processor.model.traits.Parent;

import java.io.IOException;
import java.util.*;

import static com.azure.tools.apiview.processor.analysers.models.Constants.*;

public class ReviewLine implements Parent, JsonSerializable<ReviewLine> {
    /*
     * lineId is only required if we need to support commenting on a line that contains this token.
     * Usually code line for documentation or just punctuation is not required to have lineId. lineId should be a unique
     * value within the review token file to use it assign to review comments as well as navigation Id within the review
     * page, for e.g Azure.Core.HttpHeader.Common, azure.template.template_main
     */
    private final String lineId;

    private String crossLanguageId;

    /*
     * list of tokens that constructs a line in API review.
     */
    private final List<ReviewToken> tokens;

    private final Parent parent;

    /*
     * Add any child lines as children. For e.g. all classes and namespace level methods are added as a children of
     * namespace(module) level code line. Similarly all method level code lines are added as children of it's class code
     * line.
     */
    private final List<ReviewLine> children;

    /*
     * Set current line as hidden code line by default. .NET has hidden APIs and architects don't want to see them by default.
     */
    private boolean isHidden = false;

    /*
     * Set current line as context end line. For e.g. line with token } or empty line after the class to mark end of context.
     */
    private boolean isContextEndLine;

    /*
     * Set ID of related line to ensure current line is not visible when a related line is hidden. One e.g. is a code
     * line for class attribute should set class line's Line ID as related line ID.
     */
    private String relatedToLine;

    /*
     * This is not used by APIView - this is for use by the parser only, so that it may write more diagnostics
     */
    private final Map<String, String> properties;

    public ReviewLine(Parent parent) {
        this(parent, null);
    }

    public ReviewLine(final Parent parent, final String lineId) {
        this.parent = Objects.requireNonNull(parent);
        this.tokens = new ArrayList<>();
        this.children = new ArrayList<>();
        this.lineId = lineId;
        this.properties = new HashMap<>();
    }

    public ReviewLine addToken(ReviewToken token) {
        tokens.add(token);
        return this;
    }

    public ReviewLine addToken(TokenKind tokenKind, String value) {
        return addToken(tokenKind, value, Spacing.DEFAULT);
    }

    public ReviewLine addToken(TokenKind tokenKind, String value, Spacing spacing) {
        return addToken(tokenKind, value, null, spacing);
    }

    public ReviewLine addToken(TokenKind tokenKind, String value, String navigateToId) {
        return addToken(tokenKind, value, navigateToId, Spacing.DEFAULT);
    }

    public ReviewLine addToken(TokenKind tokenKind, String value, String navigateToId, Spacing spacing) {
        return addToken(new ReviewToken(tokenKind, value, navigateToId).setSpacing(spacing));
    }

    public ReviewLine addSpace() {
        return addToken(TokenKind.TEXT, " ", Spacing.NO_SPACE);
    }

    /**
     * Returns the child that was passed in as the argument
     * @param child
     * @return
     */
    @Override
    public ReviewLine addChildLine(ReviewLine child) {
        children.add(child);
        return child;
    }

    /**
     * Add a review line with the given line ID
     * @return the new review line
     */
    @Override
    public ReviewLine addChildLine(final String lineId) {
        return addChildLine(new ReviewLine(this, lineId));
    }

    @Override
    public ReviewLine addChildLine() {
        return addChildLine(new ReviewLine(this, null));
    }

    @Override
    public List<ReviewLine> getChildren() {
        return children;
    }

    public ReviewLine addProperty(String key, String value) {
        properties.put(key, value);
        return this;
    }

    public boolean hasProperty(String key) {
        return properties.containsKey(key);
    }

    public String getProperty(String key) {
        return properties.get(key);
    }

    public List<ReviewToken> getTokens() {
        return tokens;
    }

    public ReviewLine addContextStartTokens() {
        addToken(TokenKind.PUNCTUATION, "{");
        return this;
    }

    public void addContextEndTokens() {
        parent.addChildLine(new ReviewLine(this).addToken(TokenKind.PUNCTUATION, "}").setContextEndLine());
    }

    public ReviewLine setContextEndLine() {
        isContextEndLine = true;
        return this;
    }

    public String getLineId() {
        return lineId;
    }

    public ReviewLine setRelatedToLine(ReviewLine relatedToLine) {
        this.relatedToLine = relatedToLine.lineId;
        return this;
    }

    public ReviewLine setRelatedToLine(String relatedToLineId) {
        this.relatedToLine = relatedToLineId;
        return this;
    }

    public ReviewLine setCrossLanguageId(String crossLanguageId) {
        this.crossLanguageId = crossLanguageId;
        return this;
    }

    /**
     * This line will not appear in the APIView navigation or content area.
     */
    public ReviewLine hideLine() {
        isHidden = true;
        return this;
    }

    @Override
    public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
        jsonWriter.writeStartObject();

        if (lineId != null) {
            jsonWriter.writeStringField(JSON_LINE_ID, lineId);
        }

        if (crossLanguageId != null) {
            jsonWriter.writeStringField(JSON_CROSS_LANGUAGE_ID, crossLanguageId);
        }

        if (!tokens.isEmpty()) {
            jsonWriter.writeArrayField(JSON_NAME_TOKENS, tokens, JsonWriter::writeJson);
        }

        if (!children.isEmpty()) {
            jsonWriter.writeArrayField(JSON_NAME_CHILDREN, children, JsonWriter::writeJson);
        }

        if (isHidden) {
            jsonWriter.writeBooleanField(JSON_IS_HIDDEN, isHidden);
        }

        if (isContextEndLine) {
            jsonWriter.writeBooleanField(JSON_IS_CONTEXT_END_LINE, isContextEndLine);
        }

        if (relatedToLine != null) {
            jsonWriter.writeStringField(JSON_RELATED_TO_LINE, relatedToLine);
        }

        return jsonWriter.writeEndObject();
    }
}
